Make site-isolated frames focusable

charliewolfe opened the pull request:
#### 2bfb21d47f33c8af3f8419ada9ba103a5e0b9d30
<pre>
Make site-isolated frames focusable
<a href="https://bugs.webkit.org/show_bug.cgi?id=261845">https://bugs.webkit.org/show_bug.cgi?id=261845</a>
rdar://115807972

Reviewed by Alex Christensen.

This change makes it possible to focus site-isolated frames using `.focus()`. Also broadcast the focused
frame to make sure the focused frame is always up to date in each web process and the UI process.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::trySuspendPage):
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::setFocus):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::focusedFrameChanged):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setFocusedFrame):
(WebCore::FocusController::focusedOrMainFrame const):
(WebCore::FocusController::setFocusedInternal):
(WebCore::FocusController::setFocusedElement):
(WebCore::FocusController::setActiveInternal):
(WebCore::FocusController::findFocusCandidateInContainer):
(WebCore::FocusController::advanceFocusDirectionally):
* Source/WebCore/page/FocusController.h:
(WebCore::FocusController::updateFocusedFrame):
(WebCore::FocusController::focusedFrame const):
(WebCore::FocusController::focusedLocalFrame const):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::focus):
* Source/WebKit/Shared/FrameInfoData.h:
* Source/WebKit/Shared/FrameInfoData.serialization.in:
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _isFocused]):
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfoPrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::promptForDisplayCapturePermission):
(WebKit::UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::isFocused const):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedFrameChanged):
(WebKit::WebPageProxy::broadcastFocusedFrameToOtherProcesses):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::focusedFrameChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const):
(WebKit::WebFrame::isFocused const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::handleAcceptedCandidate):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::focusedFrameChanged):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268570@main">https://commits.webkit.org/268570@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c447485511b56e8f710557f09e3aa76ea3bbf59f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20494 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21954 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18721 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20285 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23737 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20652 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20202 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22804 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18226 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24484 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22474 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18996 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18192 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22536 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2466 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18822 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1337,4 +1341,37 @@ HTTPServer server({
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://domain1.com/mainframe"]]];
+    EXPECT_WK_STREQ([webView _test_waitForAlert], "focused");
+
+    auto mainframe = WKPageGetMainFrame([webView _pageForTesting]);
```
This is currently only available on macOS.  Could we instead add SPI to WKFrameInfo?  BOOL _focused

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -439,16 +439,17 @@ FocusController::FocusController(Page& page, OptionSet<ActivityState> activitySt
 {
     ASSERT(!frame || frame->page() == &m_page);
+    m_page.chrome().focusedFrameChanged(frame);
     if (m_focusedFrame == frame || m_isChangingFocusedFrame)
```
I don't think I understand why this doesn't work with site isolation.  Could you clarify more in the commit log?

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKFrameInfoPrivate.h:
```diff
@@ -33,5 +33,6 @@
 @property (nonatomic, readonly, copy, nullable) _WKFrameHandle *_parentFrameHandle WK_API_AVAILABLE(macos(11.0), ios(14.0));
 @property (nonatomic, readonly) pid_t _processIdentifier WK_API_AVAILABLE(macos(13.3), ios(16.4));
 @property (nonatomic, readonly) BOOL _isLocalFrame WK_API_AVAILABLE(macos(14.0), ios(17.0));
+@property (nonatomic, readonly) BOOL _focused WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
_isFocused

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1337,4 +1337,51 @@ HTTPServer server({
+    Util::run(&done);
+
+    EXPECT_NOT_NULL(childFrameInfo);
+    EXPECT_TRUE(childFrameInfo.get()._focused);
```
Make sure the main frame is not focused.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1337,4 +1337,51 @@ HTTPServer server({
+    Util::run(&done);
+
+    EXPECT_NOT_NULL(mainFrameInfo);
+    EXPECT_TRUE(mainFrameInfo.get()._focused);
```
Make sure the child frame is not focused.

achristensen07 reviewed (changes requested):
iOS test needs some fixing.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1337,4 +1337,44 @@ HTTPServer server({
+    }];
+    Util::run(&done);
+
+    EXPECT_NOT_NULL(childFrameInfo);
```
These null checks aren't necessary.  But checking that the main frame is focused before the change would be good.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
```diff
@@ -109,4 +109,16 @@ - (BOOL)_isLocalFrame
+    auto focusedFrame = page->focusedFrame();
+    if (!focusedFrame)
+        return _frameInfo->isMainFrame();
+    return focusedFrame->frameID() == _frameInfo->handle()->frameID();
```
Adding a bool to FrameInfoData and getting this value from the web content process instead would allow us to check state consistency between web processes's state controllers.

I'm not sure if we also want a way to check UI process state in tests.

achristensen07 commented on Source/WebKit/Shared/FrameInfoData.h:
```diff
@@ -43,6 +43,7 @@ struct FrameInfoData {
     WebCore::FrameIdentifier frameID;
     std::optional<WebCore::FrameIdentifier> parentFrameID;
     ProcessID processID;
+    bool isFocused;
```
add { false } to prevent uninitialized memory.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1374,4 +1374,42 @@ HTTPServer server({
+    }];
+    Util::run(&done);
+
+    EXPECT_FALSE(mainFrameInfo.get()._isFocused);
```
Verify that these are switched before calling JS and that they switch back if window.focus() is called in the main frame.

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -1114,7 +1116,7 @@ static void updateFocusCandidateIfNeeded(FocusDirection direction, const FocusCa
 void FocusController::findFocusCandidateInContainer(Node& container, const LayoutRect& startingRect, FocusDirection direction, KeyboardEvent* event, FocusCandidate& closest)
 {
-    Node* focusedNode = (focusedFrame() && focusedFrame()->document()) ? focusedFrame()->document()->focusedElement() : 0;
+    Node* focusedNode = (focusedLocalFrame() && focusedLocalFrame()->document()) ? focusedLocalFrame()->document()->focusedElement() : 0;
```
nit: nullptr instead of 0 at the end.  That must be really old code.

achristensen07 commented on Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
```diff
@@ -1206,7 +1206,7 @@ static _WKAutoplayEvent toWKAutoplayEvent(WebCore::AutoplayEvent event)
     if (auto* mainFrame = frame.page() ? frame.page()->mainFrame() : nullptr)
         mainFrameID = mainFrame->frameID();
-    FrameInfoData frameInfo { frame.isMainFrame(), FrameType::Local, { }, userMediaOrigin.securityOrigin(), { }, frame.frameID(), mainFrameID, frame.process().processID() };
+    FrameInfoData frameInfo { frame.isMainFrame(), FrameType::Local, { }, userMediaOrigin.securityOrigin(), { }, frame.frameID(), mainFrameID, frame.process().processID(), frame.isMainFrame() };
```
frame.isMainFrame() isn't correct.  I'd instead add a WebFrameProxy::isFocused() that returns true if its page is non-null and if its page's focusedFrame() is equal to it.  Then use that 3x in this PR.

achristensen07 commented on Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
```diff
@@ -103,7 +103,8 @@ void WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction(const Navigat
         WTFMove(parentFrameID),
-        getCurrentProcessID()
+        getCurrentProcessID(),
+        navigationAction.initiatedByMainFrame() == InitiatedByMainFrame::Yes
```
This should do the same thing as WebFrame::info, but with requestingFrame if it's non-null.  Make WebFrame::isFocused() to share code in the 2 locations.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1304,4 +1304,63 @@ HTTPServer server({
+            mainFrameInfo = mainFrame.info;
+            childFrameInfo = mainFrame.childFrames.firstObject.info;
+        }];
+        Util::runFor(0.1_s);
```
This is a little strange.  I'd expect Util::run(&done); here and set done to true once the _frames block is called.  Or even better, a shared lambda like this:

auto getUpdatedFrameInfo = [&] {
    done = false;
    [webView _frames:^(_WKFrameTreeNode *mainFrame) {
        mainFrameInfo = mainFrame.info;
        childFrameInfo = mainFrame.childFrames.firstObject.info;
        done = true;
    }];
    Util::run(&done);
};

do {
    getUpdatedFrameInfo();
} while (!mainFrameInfo.get()._isFocused || childFrameInfo.get()._isFocused); // Also, this should be || instead of &&

evaluateJavaScript

another do loop

charliewolfe committed:
Make site-isolated frames focusable
https://bugs.webkit.org/show_bug.cgi?id=261845
rdar://115807972

Reviewed by Alex Christensen.

This change makes it possible to focus site-isolated frames using `.focus()`. Also broadcast the focused
frame to make sure the focused frame is always up to date in each web process and the UI process.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::FrameSelection):
* Source/WebCore/history/BackForwardCache.cpp:
(WebCore::BackForwardCache::trySuspendPage):
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::setFocus):
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::focusedFrameChanged):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setFocusedFrame):
(WebCore::FocusController::focusedOrMainFrame const):
(WebCore::FocusController::setFocusedInternal):
(WebCore::FocusController::setFocusedElement):
(WebCore::FocusController::setActiveInternal):
(WebCore::FocusController::findFocusCandidateInContainer):
(WebCore::FocusController::advanceFocusDirectionally):
* Source/WebCore/page/FocusController.h:
(WebCore::FocusController::updateFocusedFrame):
(WebCore::FocusController::focusedFrame const):
(WebCore::FocusController::focusedLocalFrame const):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::focus):
* Source/WebKit/Shared/FrameInfoData.h:
* Source/WebKit/Shared/FrameInfoData.serialization.in:
* Source/WebKit/UIProcess/API/APIFrameInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfo.mm:
(-[WKFrameInfo _isFocused]):
* Source/WebKit/UIProcess/API/Cocoa/WKFrameInfoPrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::promptForDisplayCapturePermission):
(WebKit::UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::cancel):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::isFocused const):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedFrameChanged):
(WebKit::WebPageProxy::broadcastFocusedFrameToOtherProcesses):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::focusedFrameChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::info const):
(WebKit::WebFrame::isFocused const):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::handleAcceptedCandidate):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::focusedFrameChanged):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268570@main
